include/lsst/partition/ConfigStore.h

```diff
@@ -0,0 +1,290 @@
+                throw lsst::partition::ConfigTypeError(
+                    context + "parameter: '" + path + "' is a string, but not the single-character one"
+                    " to allow interpreting it as a value of the 'char' type.");
+            }
+            return s[0];
+        }
+        try {
+            return param.get<char>();
+        } catch(std::exception const& ex) {
+            throw lsst::partition::ConfigTypeError(
+                context + "incorrect type conversion for parameter: '" + path + "', details: "
+                + std::string(ex.what()));
+        }
+    }
+};
+}}} // namespace lsst::partition::detail
+
+
+namespace lsst {
+namespace partition {
+/**
+ * Class ConfigStore is a unified transient storage of the configuration parameters read
+ * from the configuration files or command-line parameters.
+ * 
+ * Parameters are fetched from the store using their "path" specification. The dot
+ * character '.' is used for separating components of the path:
+ * @code
+ * foo.bar.p1
+ * dir.p2
+ * p3
+ * @code
+ * The intermediate elements of the path would directly map to the corresponding keys
+ * of the nested JSON objects in the store. For example, the above shown paths would
+ * be represented by the following JSON structure passed into the class's constructr,
+ * into method add(), or read from an input file when calling method parse().
+ * from the :
+ * @code
+ * {
+ *   "foo": {
+ *       "bar": {
+ *           "p1": <value>
+ *       }
+ *    },
+ *    "dir": {
+ *       "p2": <value>
+ *    },
+ *    "p3": <value>
+ * }
+ * @code
+ * Values of the command-line parameters are expected to be injected into the configuration
+ * by calling method set(). The long names (excluding preceeding "--") of the parameters
+ * would be treated as above explained path names.
+ *
+ * @note the class's implementation is not thread-safe, and any subsequent modification
+ *   to a parameter would always override its previous value (if any existed).
+ */
+class ConfigStore {
+public:
+    /**
+     * Initialize the store with parameters found in the input JSON object.
+     *
+     * Empty JSON objects are allowed on the input.
+     *
+     * @param config  A JSON object to be used to initialize the store.
+     * @throw std::invalid_argument  If the input value is not a JSON Object.
+     */
+    explicit ConfigStore(nlohmann::json const& config=nlohmann::json::object());
+
+    /**
+     * Parse the content of the JSON file and merge parameters found in the file into
+     * the object.
+     * 
+     * It's allowed to call this method many times for the same or different files.
+     * Parameters read from a file will be merged into the store.
+     *
+     * @param filename  The name of a JSON file to be parsed.
+     * @throw std::invalid_argument  The file name is not valid.
+     * @throw std::runtime_error  Failed to open or parse the file.
+     */
+    void parse(std::string const& filename);
```

Comment:
I'd consider naming this as an overload of the `add` function, since it parses and adds. (the other two add functions also do a read-and-add type of behavior).

Reply:
In general, this is good suggestion. Unfortunately, it wouldn't work here since `nlohman::json` seems to have a non-explicit constructor from a string. This cases the following compilation problem:
```
tests/configStore.cc: In member function 'void ConfigStoreTest::test_method()':
tests/configStore.cc:325:29: error: call of overloaded 'add(const char [1])' is ambiguous
         anotherConfig.add("");
                             ^
In file included from tests/configStore.cc:25:0:
include/lsst/partition/ConfigStore.h:188:10: note: candidate: void lsst::partition::ConfigStore::add(const string&)
     void add(std::string const& filename);
          ^~~
include/lsst/partition/ConfigStore.h:200:10: note: candidate: void lsst::partition::ConfigStore::add(const json&)
     void add(nlohmann::json const& config);
          ^~~
```
So, I'm reversing it back `::parse`.